# SVC Model

In [49]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in /Users/bunmiolakanmi/opt/anaconda3/lib/python3.8/site-packages (0.0)


In [50]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [51]:
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [52]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna() 
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [53]:
df.columns

Index(['koi_disposition', 'koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co',
       'koi_fpflag_ec', 'koi_period', 'koi_period_err1', 'koi_period_err2',
       'koi_time0bk', 'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact',
       'koi_impact_err1', 'koi_impact_err2', 'koi_duration',
       'koi_duration_err1', 'koi_duration_err2', 'koi_depth', 'koi_depth_err1',
       'koi_depth_err2', 'koi_prad', 'koi_prad_err1', 'koi_prad_err2',
       'koi_teq', 'koi_insol', 'koi_insol_err1', 'koi_insol_err2',
       'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_steff_err1',
       'koi_steff_err2', 'koi_slogg', 'koi_slogg_err1', 'koi_slogg_err2',
       'koi_srad', 'koi_srad_err1', 'koi_srad_err2', 'ra', 'dec',
       'koi_kepmag'],
      dtype='object')

In [54]:
# import os
# os.getcwd()

# Select your features (columns)

In [55]:
# Set features (columns). We are practically eliminating all error columns, and the target column, here. This will also be used as your x values.
selected_features = df[['koi_period',  'koi_time0bk', 'koi_depth', 'koi_prad', 'koi_teq', 'koi_insol', 'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_slogg', 'koi_srad', 'ra', 'dec', 'koi_kepmag']]
selected_features.head()

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
0   54.418383   162.513840      874.8      2.83      443       9.11   
1   19.899140   175.850252    10829.0     14.60      638      39.30   
2    1.736952   170.307565     8079.2     33.46     1395     891.96   
3    2.525592   171.595550      603.3      2.75     1406     926.16   
4    4.134435   172.979370      686.0      2.77     1160     427.65   

   koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad         ra  \
0           25.8                 2       5455      4.467     0.927  291.93423   
1           76.3                 1       5853      4.544     0.868  297.00482   
2          505.6                 1       5805      4.564     0.791  285.53461   
3           40.9                 1       6031      4.438     1.046  288.75488   
4           40.2                 2       6046      4.486     0.972  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714

# Create a Train Test Split

Use `koi_disposition` for the y values

In [56]:
X=selected_features
X

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
0      54.418383   162.513840      874.8      2.83      443       9.11   
1      19.899140   175.850252    10829.0     14.60      638      39.30   
2       1.736952   170.307565     8079.2     33.46     1395     891.96   
3       2.525592   171.595550      603.3      2.75     1406     926.16   
4       4.134435   172.979370      686.0      2.77     1160     427.65   
...          ...          ...        ...       ...      ...        ...   
6986    8.589871   132.016100       87.7      1.11      929     176.40   
6987    0.527699   131.705093     1579.2     29.35     2088    4500.53   
6988    1.739849   133.001270       48.5      0.72     1608    1585.81   
6989    0.681402   132.181750      103.6      1.07     2218    5713.41   
6990    4.856035   135.993300       76.7      1.05     1266     607.42   

      koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad  \
0              25.8                 2       5455      4.467     0.927   
1              76.3                 1       5853      4.544     0.868   
2             505.6                 1       5805      4.564     0.791   
3              40.9                 1       6031      4.438     1.046   
4              40.2                 2       6046      4.486     0.972   
...             ...               ...        ...        ...       ...   
6986            8.4                 1       5638      4.296     1.088   
6987          453.3                 1       5638      4.529     0.903   
6988           10.6                 1       6119      4.444     1.031   
6989           12.3                 1       6173      4.447     1.041   
6990            8.2                 1       6469      4.385     1.193   

             ra        dec  koi_kepmag  
0     291.93423  48.141651      15.347  
1     297.00482  48.134129      15.436  
2     285.53461  48.285210      15.597  
3     288.75488  48.226200      15.509  
4     296.28613  48.224670      15.714  
...         ...        ...         ...  
6986  298.74921  46.973351      14.478  
6987  297.18875  47.093819      14.082  
6988  286.50937  47.163219      14.757  
6989  294.16489  47.176281      15.385  
6990  297.00977  47.121021      14.826  

[6991 rows x 14 columns]

In [57]:
y = df["koi_disposition"]
target_names = ['CONFIRMED', 'FALSE POSITIVE', 'CANDIDATE']

In [58]:
# Split data to train and test datasets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [59]:
X_train.head()

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
6122    6.768901   133.077240      123.1      1.24     1017     253.30   
6370    0.733726   132.020050      114.6      0.86     1867    2891.64   
2879    7.652707   134.460380      641.1      3.21      989     226.81   
107     7.953547   174.662240      875.4      2.25      696      55.37   
29      4.959319   172.258529     9802.0     12.21     1103     349.40   

      koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad  \
6122           10.8                 1       5737      4.327     1.125   
6370           13.8                 1       5855      4.578     0.797   
2879          254.3                 1       6328      4.481     0.963   
107            38.4                 1       4768      4.536     0.779   
29            696.5                 1       5712      4.359     1.082   

             ra        dec  koi_kepmag  
6122  294.40472  39.351681      14.725  
6370  284.50391  42.463860      15.770  
2879  295.50211  38.983540      13.099  
107   291.15878  40.750271      15.660  
29    292.16705  48.727589      15.263

In [60]:
y_train.head()

6122         CANDIDATE
6370    FALSE POSITIVE
2879    FALSE POSITIVE
107          CONFIRMED
29           CANDIDATE
Name: koi_disposition, dtype: object

# Pre-processing

    -Scale the data using the StandardScaler 
    -Using `MinMaxScaler` results in model with lower performance

In [100]:
# Scale your data
#from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
X_scaler = StandardScaler().fit(X_train)

In [101]:
# Transform the training and testing data using the X_scaler and y_scaler models
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [102]:
# from sklearn.svm import LinearSVC
# # from sklearn.datasets import load_iris
# from sklearn.feature_selection import SelectFromModel
# # X, y = load_iris(return_X_y=True)
# # X.shape
# # (150, 4)
# lsvc = LinearSVC(C=0.01, penalty="l1", dual=False).fit(X_train_scaled, y_train)
# model = SelectFromModel(lsvc, prefit=True)
# X_new = model.transform(X_train_scaled)
# X_new.shape
# # (150, 3)
# model.get_support()
# selected_feat= X_train.columns[(model.get_support())]
# selected_feat

# Train the Model



In [103]:
# Support vector machine linear classifier
from sklearn.svm import SVC 
model = SVC(kernel='linear')
model.fit(X_train_scaled, y_train)

SVC(kernel='linear')

In [104]:
# Model Accuracy
print('Test Acc: %.3f' % model.score(X_test_scaled, y_test))

Test Acc: 0.614


In [105]:
# Calculate classification report
from sklearn.metrics import classification_report
predictions = model.predict(X_test_scaled)
print(classification_report(y_test, predictions,
                            target_names=target_names))

                precision    recall  f1-score   support

     CONFIRMED       0.38      0.25      0.30       411
FALSE POSITIVE       0.56      0.60      0.58       484
     CANDIDATE       0.71      0.80      0.75       853

      accuracy                           0.61      1748
     macro avg       0.55      0.55      0.54      1748
  weighted avg       0.59      0.61      0.60      1748



In [106]:
print(f"Training Data Score: {model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model.score(X_test_scaled, y_test)}")

Training Data Score: 0.6395193591455274
Testing Data Score: 0.6138443935926774


# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [107]:
# Create the GridSearch estimator along with a parameter object containing the values to adjust
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1, 5, 10],
              'gamma': [0.0001, 0.001, 0.01]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [108]:
# param_grid = {'C': [0.1, 1, 10, 100, 1000], 
#               'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
#               'kernel': ['rbf']} 

In [109]:
# # Train the model with GridSearch
# grid = GridSearchCV(SVC(), param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................... C=1, gamma=0.0001, score=0.630, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.632, total=   0.4s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.9s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.631, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.625, total=   0.4s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.636, total=   0.5s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.630, total=   0.5s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.632, total=   0.5s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.631, total=   0.5s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.625, total=   0.4s
[CV] C=1, gamma=0.001 ................................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:   37.8s finished


GridSearchCV(estimator=SVC(kernel='linear'),
             param_grid={'C': [1, 5, 10], 'gamma': [0.0001, 0.001, 0.01]},
             verbose=3)

In [110]:
# print best parameter after tuning
print(grid.best_params_)
  
# print how our model looks after hyper-parameter tuning
print(grid.best_estimator_)

{'C': 10, 'gamma': 0.0001}
SVC(C=10, gamma=0.0001, kernel='linear')


In [111]:
# print(grid2.best_params_)
print(grid.best_score_)

0.6385649000502115


In [112]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [113]:
print('Test Acc: %.3f' % grid.score(X_test_scaled, y_test))

Test Acc: 0.629


In [114]:
# Calculate classification report
from sklearn.metrics import classification_report
predictions = model.predict(X_test_scaled)
print(classification_report(y_test, predictions,
                            target_names=target_names))

                precision    recall  f1-score   support

     CONFIRMED       0.38      0.25      0.30       411
FALSE POSITIVE       0.56      0.60      0.58       484
     CANDIDATE       0.71      0.80      0.75       853

      accuracy                           0.61      1748
     macro avg       0.55      0.55      0.54      1748
  weighted avg       0.59      0.61      0.60      1748



In [115]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions,
                            target_names))

                precision    recall  f1-score   support

     CONFIRMED       0.56      0.60      0.58       484
FALSE POSITIVE       0.71      0.80      0.75       853
     CANDIDATE       0.38      0.25      0.30       411

      accuracy                           0.61      1748
     macro avg       0.55      0.55      0.54      1748
  weighted avg       0.59      0.61      0.60      1748



/Users/bunmiolakanmi/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:67: FutureWarning: Pass labels=['CONFIRMED', 'FALSE POSITIVE', 'CANDIDATE'] as keyword args. From version 0.25 passing these as positional arguments will result in an error
  warnings.warn("Pass {} as keyword args. From version 0.25 "


# Save the Model

In [116]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'SVC_grid.sav'
joblib.dump(grid, filename)

['SVC_grid.sav']

#### Inference(s) 
* The svc model performance after tuning (0.629) is just a little bit better than its performance before tuning (0.614).